packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/bitmap.dart

```diff
@@ -175,6 +230,116 @@ class BitmapDescriptor {
+
+    final double devicePixelRatio =
+        WidgetsBinding.instance.platformDispatcher.views.first.devicePixelRatio;
+    final double? targetImagePixelRatio =
+        imagePixelRatio ?? configuration.devicePixelRatio;
+    final Size? size = configuration.size;
+
+    if (!mipmaps && (targetImagePixelRatio != null || size != null)) {
+      return BitmapDescriptor._(<Object>[
+        _asset,
+        assetName,
+        _getBitmapScalingString(bitmapScaling),
+        targetImagePixelRatio ?? devicePixelRatio,
+        if (size != null)
+          <Object>[
+            size.width,
+            size.height,
+          ],
+      ]);
+    }
+
+    final AssetImage assetImage =
+        AssetImage(assetName, package: package, bundle: bundle);
+    final AssetBundleImageKey assetBundleImageKey =
+        await assetImage.obtainKey(configuration);
+
+    return BitmapDescriptor._(<Object>[
+      _asset,
+      assetBundleImageKey.name,
+      _getBitmapScalingString(bitmapScaling),
+      assetBundleImageKey.scale,
+      if (size != null)
+        <Object>[
+          size.width,
+          size.height,
+        ],
+    ]);
+  }
+
+  /// Creates a BitmapDescriptor using an array of bytes that must be encoded
+  /// as PNG.
+  ///
+  /// The optional [size] parameter represents the *logical size* of the
+  /// bitmap, regardless of the actual resolution of the encoded PNG.
+  /// This helps the platform to render High-DPI images at the correct size.
+  /// [imagePixelRatio] value can be use to scale the image to
+  /// proper size across platforms.
+  ///
+  /// Note: Image scaling with [size] parameter does not maintain the aspect ratio.
+  /// If a proportional resize is required, ensure to provide a [size] with the
+  /// correct aspect ratio.
+  static BitmapDescriptor createFromBytes(
+    Uint8List byteData, {
+    BitmapScaling bitmapScaling = BitmapScaling.auto,
+    double? imagePixelRatio,
+    Size? size,
+  }) {
+    assert(byteData.isNotEmpty,
+        'Cannot create BitmapDescriptor with empty byteData.');
+    assert(bitmapScaling != BitmapScaling.noScaling || imagePixelRatio == null,
+        'If bitmapScaling is set to BitmapScaling.noScaling, scale parameter cannot be used.');
+    assert(bitmapScaling != BitmapScaling.noScaling || size == null,
+        'If bitmapScaling is set to BitmapScaling.noScaling, size parameter cannot be used.');
+
+    return BitmapDescriptor._(<Object>[
+      _bytes,
+      byteData,
+      _getBitmapScalingString(bitmapScaling),
+      imagePixelRatio ?? 1.0,
+      if (size != null)
+        <Object>[
+          size.width,
+          size.height,
+        ],
+    ]);
+  }
+
+  static String _getBitmapScalingString(BitmapScaling bitmapScaling) {
+    switch (bitmapScaling) {
+      case BitmapScaling.auto:
+        return bitmapAutoScaling;
+      case BitmapScaling.noScaling:
+        return bitmapNoScaling;
+    }
+  }
+
   final Object _json;
```

Comment:
These _json-backed objects are ugly, and force us to write ugly code in the `convert.dart`.

Is there any way that this could be separated into something like:

`abstract class MapBitmap extends BitmapDescriptor` (so it can be passed around) and then have:

`class AssetMapBitmap extends MapBitmap` and `class BytesMapBitmap extends MapBitmap` which can be created as usual?

Then the changes in the convert.dart and elsewhere would be minimal; just a type check and a cast to use the right subtype and its accessors?

It seems the only public method on all of these are `.toJson`, which can probably be skipped for the new classes.

(Not sure how complex this is for other platforms, but I suspect that the "json" object is "convenient" for the method channel serialization?)